General Scaling Up document (#48)
OSGConnect/connectbook: start/scaling-up/scaling-up.md

[title]: - "Scale Up your Workflow on OSG Connect"
[TOC]

## Background

Much of OGS's computing power comes from the ability to run a large number 
of jobs simulateously. Many HTC-friendly workflows involve running the same 
job multiple times with no or minor changes between them. Here, we will 
talk about how to submit multiple jobs at once using the `queue` command as 
well as some important considerations and tips for scaling up your HTCondor jobs. 
For a hands-on example of these concepts, checkout the [OSG Connect Quickstart]
(https://support.opensciencegrid.org/support/solutions/articles/5000633410-osg-connect-quickstart). 

> ## Things to Consider

Reviewer: Maybe a different title? I think of "Things to Consider" as more about the actual jobs, not the terminology.

Reviewer: oh, never mind, I wasn't looking at the header level carefully.

In order to avoid confusion, let's review some HTCondor terminiology:
 - When you queue a single submit script using the `condor_submit` command, this is 

> called a **cluster** and can consist of one more individual jobs.

Reviewer: ```suggestion
called a **cluster** and can consist of one or more individual jobs.

```

 - Each individual job within a cluster is called a **process**. Therefore, each 

> *cluster* can be made up of one or more *processees*. Each proceess lands on its 

Reviewer: ```suggestion
*cluster* can be made up of one or more *processes*. Each process lands on its 

```

own job slot and is handled individually from the other processes in the cluster.

> ### Handling Output

Reviewer: Make this handling input + output?  This guide doesn't yet cover how to do something like unique input files / arguments.

Author: As below: Added a section on Submitting Multiple Jobs with a Single Script which mentions this option and links to CHTC's doc page on it.

HTCondor has built in features that make submitting multiple jobs with the same 
submit script possible and easy. However, this means that you will need to ensure 

> that the output created by each individual process will be uniquely identified so 

Reviewer: ```suggestion
that the output created by each individual job will be uniquely identified so 

```

Reviewer 2: why use 'job' here and not 'process'?

Reviewer: I guess either is fine. Process is an overloaded term that has other meanings, so I wanted to make it clear that we're talking about a job here, not a process within a larger program.

Author: I went with process to keep terms straight with the assumption that the reader would use the definition above and that using job would add confusion.

Author: I could italicize it to further emphasize that I'm using the same term as defined.

Reviewer: I think it should be fine. You can leave it as is. :)

it is not overwritten by the output from other processes.

In order to uniquely identify each job output, you can use cluster and process IDs
in the submit script. For example, to route each job log into it's own file, the 
following line would ensure that each job's log file would end up with a unique file 
name:

	log = job.$(Cluster).$(Process).log

For a working example on how to use this in a submit file, and additional applications,
please see [Job 3 of the OSG Connect Quickstart guide]

> (https://support.opensciencegrid.org/support/solutions/articles/5000633410-osg-connect-quickstart#job-3-submitting-jobs-concurrently)

Reviewer: It would also be nice to refer to examples where users provide a list of values or use a glob instead of using the `$(Process)` numbers. If we don't show in detail anywhere in our guides (yet), maybe refer to HTCondor Week talks?

Author: Added a section on Submitting Multiple Jobs with a Single Script which mentions this option and links to CHTC's doc page on it.

### Disk Usage

> Each OSG Connect user is granted 500 GB of storage in their `/public` directories. 

Reviewer: add something about `/home` also?
